### ISSUE TITLE
Remaining amount of MetaDEx trade is not updated

### ISSUE THREAD

**Issue by dexX7** [OmniLayer/omnicore#200]
When using `"omni_gettrade"`, even though there were matches, then the "dynamic" values of the trade are not updated:
```js
{
    "txid" : "e08017247b0096b20cd912fa61833d41d37a5dd38e6fc13b7797774085d0755b",
    "fee" : "0.00010000",
    "sendingaddress" : "mrPJfUmbeh6bzZ7zWqmgsjQSa4fP5NnY9F",
    "ismine" : true,
    "version" : 0,
    "type_int" : 25,
    "type" : "MetaDEx trade",
    "propertyidforsale" : 1,
    "propertyidforsaleisdivisible" : true,
    "amountforsale" : "20.00000000",
    "amountremaining" : "20.00000000", // <---- already 3.0 sold!
    "propertyiddesired" : 5,
    "propertyiddesiredisdivisible" : false,
    "amountdesired" : "10",
    "amounttofill" : "10",
    "unitprice" : "2.00000000000000000000000000000000000000000000000000",
    "status" : "open part filled",
    "matches" : [
        {
            "txid" : "17edfe958b936060638672e71583b1ccf2f825dbcd6a9a7a0ae432b583355d8a",
            "block" : 180,
            "address" : "mrPJfUmbeh6bzZ7zWqmgsjQSa4fP5NnY9F",
            "amountsold" : "2.00000000",
            "amountreceived" : "1"
        },
        {
            "txid" : "f48e650f146c632a3e85cd3ef3250856c91785596b681b032461dff1f9a57392",
            "block" : 179,
            "address" : "mqy4X6JdLWMCz1eGDAfPCZLKr13qYPBNnh",
            "amountsold" : "1.00000000",
            "amountreceived" : "1"
        }
    ],
    "valid" : true,
    "blockhash" : "4dffb9e40d2dee0c7fc838428d48ecee04e312bb4ffdac5061e5da44982dd97c",
    "blocktime" : 1440803439,
    "block" : 179,
    "confirmations" : 2
}
```

This is because we currently handle trades as stateless transactions.

The trade objects are created based on `CMPMetaDEx` via the (parsed) `CMPTransaction`.

It may be resolved by looking up the actual trade object from the in-memory storage, or alternatively try to reconstruct the trade. This isn't really great though, and unfortunally we only keep active trades in memory.

**Comment by zathras-crypto** [comment]
Perhaps semantics but technically there is nothing available (remaining) if the trade is closed (filled or cancelled).

Thus since ```amountremaining``` and ```amounttofill``` are only relevant for open trades we could take the simple path and only provide those fields if we can obtain the relevant data from a trade in-memory storage (ie it's still open).

Alternatively we could go through a more complex process and provide the field at all times (so we're effectively providing data on ```amountremaining``` *at the time the trade was cancelled* or a zero value if the trade was filled) but this seems like it could cause confusion (ie why would there be an ```amountremaining``` and ```amounttofill``` on a closed trade for example since it's not possible to fill it).

Summary...

Option 1: Closed trades (filled or cancelled) do not have fields relating to what's available and the cost to fill it (ie hide ```amountremaining``` and ```amounttofill```).  Open trades still have the fields and the data is pulled from the MetaDEx maps.

Option 2: All trades (including closed) have ```amountremaining``` and ```amounttofill``` fields.  For filled trades these are both set to zero, for cancelled trades some calculations are done to ascertain the amount remaining & cost to fill at the time of cancellation and those values are provided.

Option 1 seems like the simplest path and in my mind provides a more consistent data-set (since we are not saying n tokens are remaining when in fact a trade is closed).

**Comment by dexX7** [comment]
> Perhaps semantics but technically there is nothing available (remaining) if the trade is closed (filled or cancelled).

Very good point.

I already tested looping over the in-memory map, and if matched (by txid), the trade object is replaced, but it feels a bit like "retrieving the adjusted amount for sale of the traditional DEx", though it works.

Do we have this issue only with tx 25?

Alternatively, I retrieved the in-memory trade directly in the "stateless" part, but maybe we should move the "amount still for sale" and "to fill" into the extended details? If so, I'd feel more comfortable with it, because then we have a clear seperatation between stateless (i.e. "only data from the parsed transaction") vs. stateful (in-memory trade, DB access, ...).

**Comment by zathras-crypto** [comment]
> Alternatively, I retrieved the in-memory trade directly in the "stateless" part, but maybe we should move the "amount still for sale" and "to fill" into the extended details?

I agree with this - I started heading in that direction where ideally ```omni_gettransaction``` should be completely stateless (ie parse the packet and dump the data regardless of state) however this is confused a bit by the fact we have the ```valid``` attribute (which is state based).

> Do we have this issue only with tx 25?

Interesting Q, I don't have an immediate answer but my suspicion is we may have this type of issue elsewhere (either in STO or DEx phase 1)

**Comment by dexX7** [comment]
> however this is confused a bit by the fact we have the valid attribute (which is state based).

The chain information (block, confirmations, ...) and validity are only included for confirmed transactions.

**Comment by zathras-crypto** [comment]
> The chain information (block, confirmations, ...) and validity are only included for confirmed transactions.

Understood, but it's still state based - I'm speaking in the context of
> clear seperatation between stateless (i.e. "only data from the parsed transaction") vs. stateful (in-memory trade, DB access, ...).

Which I agree with, but when we're including validity (which is based on prior state) we can't limit ```omni_gettransaction``` purely to "only data from the parsed transaction"

Hope that makes sense :)

**Comment by dexX7** [comment]
Ah, now I see, yes, you're right! So we basically should have something like "populateBaseInfo" (txid, ...) and "populateBaseInfoExtended" (block, validity, ...).

> I started heading in that direction where ideally omni_gettransaction should be completely stateless

How would the user get stateful information then? We can assume the user usually wants stateful information (to some degree at least).

**Comment by zathras-crypto** [comment]
> How would the user get stateful information then?

Well that was the problem, I couldn't come up with a unified model so moved onto other things - but you can kind of see the thoughts behind it in MetaDEx:

* ```omni_gettransaction``` only shows the transaction itself (stateless)
* ```omni_gettrade``` shows the transaction and interactions (stateful)

Without removing validity from ```omni_gettransaction``` though it doesn't really work from a separation PoV, and I'm half-tempted to just flatten the entire thing and suggest moving stateful stuff into ```omni_gettransaction``` too - long story short as you've probably guessed by now I have many thoughts here but no single path to suggest :)

**Comment by dexX7** [comment]
Well, in general I like the idea of having `populateXX`, and `populateXXExtended`, and we mostly have a seperation here. It looks like we have to differentiate further, given that we want validity for example, but not necessarily all state details, so a strict seperation between "stateful" and "stateless" doesn't seem useful, and we need more "shades".

I was playing with the thoughts of adding a flag to `omni_gettransaction` to trigger the extended details, and ultimately deprechiate all transaction calls, such as `omni_gettrade`, `omni_getsto`, ... which already mostly act as forwarder. On the other hand, having very specific calls may not be bad either, though I then think we should be consistent, and provide "specific" calls for everything.

Regaring the initial issue: I think we should try to move `"amounttofill"` and `"amountremaining"` into the extended details, which should likely resolve the issue.

**Comment by dexX7** [comment]
Quick note: I tagged this with 0.0.10, because it's probably low hanging. Not sure about the other twos, so feel free to use the tag as you like. :)

**Comment by zathras-crypto** [comment]
> Quick note: I tagged this with 0.0.10, because it's probably low hanging

It was pretty striaght forward yep - see https://github.com/OmniLayer/omnicore/pull/203

**Comment by zathras-crypto** [comment]
> On the other hand, having very specific calls may not be bad either, though I then think we should be consistent, and provide "specific" calls for everything.

It's hard to find the right terminology here, but basically I tend to think of it as:

 *Is there other relevant data that is not directly expressed in the payload/message?*  

If the answer is yes, then ```omni_gettransaction``` doesn't "show the whole picture" and there is need for an extended call (```omni_getXXYYZZ``` and friends).  

If the answer is no, there is no other relevant data that isn't already in ```omni_gettransaction``` and an extended call would be pointless (eg what additional data would we put in a ```omni_getsimplesend``` call?).

I'm not sure if that'll come across clearly or not hehe :)

**Comment by dexX7** [comment]
> eg what additional data would we put in a omni_getsimplesend call?

Hehe, crowdsale participations. ;)

I mostly agree with the sentiment here. Is there anything specific what you have in mind right now, what we should do in this context?

**Comment by dexX7** [comment]
Let's continue in #204. :)

**Comment by dexX7** [comment]
This issue was resolved via #203.
